filewatcher/filewatcher.go

```diff
@@ -102,22 +127,38 @@ func (r *Result) watcherLoop(
 	softLimit, hardLimit int64,
 	logger log.Wrapper,
 	pollingInterval time.Duration,
+	fsEventsDelay time.Duration,
+	watchedFiles []string,
 ) {
-	forceReload := func(mtime time.Time) {
-		f, err := limitopen.OpenWithLimit(path, softLimit, hardLimit)
-		if err != nil {
-			logger.Log(context.Background(), "filewatcher: I/O error: "+err.Error())
-			return
-		}
-		defer f.Close()
-		d, err := parser(f)
+	var lock sync.Mutex
+	forceReload := func() {
+		// make sure we don't run forceReload concurrently
+		lock.Lock()
+		defer lock.Unlock()
+
+		d, mtime, files, err := openAndParse(path, parser, softLimit, hardLimit)
 		if err != nil {
-			logger.Log(context.Background(), "filewatcher: parser error: "+err.Error())
+			logger.Log(context.Background(), err.Error())
 		} else {
 			r.data.Store(&atomicData{
 				data:  d,
 				mtime: mtime,
 			})
+			// remove all previously watched files
+			for _, path := range watchedFiles {
```

Comment:
loop through `watcher.WatchedFiles()` or whatever instead?

Reply:
that's actually add by a fork of the original `fsnotify` package we were using, but since the original one is already stopped developing it's time to switch to the fork.

Reply:
:ohno: